File: LinkedList/Reverse_Alt_K_Nodes.py
# Python program to alternately reverse a linked list in group of given size k


# Node class
class Node:

    # Constructor to initialise data and next
    def __init__(self, data=None):
        self.data = data
        self.next = None


class SinglyLinkedList:

    # Constructor to initialise head
    def __init__(self):
        self.head = None

    # Function to reverse K nodes of linked list
    def reverse_k_nodes(self, head, k):
        current = head
        next = None
        prev = None
        count = 0

        # traverse k nodes ahead reversing the links or until current is not None
        while current is not None and count < k:
            next = current.next
            current.next = prev
            prev = current
            current = next
            count += 1

        if head is not None:
            head.next = current

        count = 0
        # traverse the k nodes to be skipped
        while current is not None and count < k-1:
            current = current.next
            count += 1

        # recursive call to the function to alternate reverse the remaining n-2k nodes
        if current is not None:
            current.next = self.reverse_k_nodes(current.next, k)

        # return the new header of the current sublist
        return prev

    # Function to Insert data at the beginning of the linked list
    def insert_at_beg(self, data):
        node = Node(data)
        node.next = self.head
        self.head = node

    # Function to print the linked list
    def print_data(self):
        current = self.head
        while current is not None:
            print(current.data, '-> ', end='')
            current = current.next
        print('None')

if __name__ == '__main__':
    linked_list = SinglyLinkedList()
    linked_list.insert_at_beg(10)
    linked_list.insert_at_beg(9)
    linked_list.insert_at_beg(8)
    linked_list.insert_at_beg(7)
    linked_list.insert_at_beg(6)
    linked_list.insert_at_beg(5)
    linked_list.insert_at_beg(4)
    linked_list.insert_at_beg(3)
    linked_list.insert_at_beg(2)
    linked_list.insert_at_beg(1)
    linked_list.print_data()
    # call the reverse k nodes function
    linked_list.head = linked_list.reverse_k_nodes(linked_list.head, 3)
    # print the reversed list
    linked_list.print_data()

File: LinkedList/Reverse_Alt_K_Nodes.cpp
// Reverse a singly linked list alternately in groups of size k

#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node* next;

        Node() {
            this->next = NULL;
        }

        Node(int data) {
            this->data = data;
            this->next = NULL;
        }
};

class LinkedList {
    public:
        Node* head = NULL;

        void insert_at_beg(int data){
            Node* temp = new Node(data);
            temp->next = this->head;
            this->head = temp;
        }

        void print_data(){
            Node* itr = this->head;

            while(itr != NULL){
                cout << itr->data << " ";
                itr = itr->next;
            }
            cout << endl;
        }
};

Node* reverse_alt_k_nodes_util(Node* head, int k) {
    // if linked list is empty or has only 1 node
    if(head == NULL || head->next == NULL) return head;

    Node *curr_ptr = head, *next_ptr, *prev_ptr = NULL;
    int count = 0;

    while(curr_ptr != NULL && count < k){
        count += 1;
        next_ptr = curr_ptr->next;
        curr_ptr->next = prev_ptr;
        prev_ptr = curr_ptr;
        curr_ptr = next_ptr;
    }
    
    head->next = curr_ptr;

    count  = 0;
    while(curr_ptr != NULL && count < k-1) {
        curr_ptr = curr_ptr->next;
        count += 1;
    }
    if (curr_ptr != NULL)    
        curr_ptr->next = reverse_alt_k_nodes_util(curr_ptr->next, k);

    // update to the new head
    return prev_ptr;
}

void reverse_alt_k_nodes(LinkedList *ll, int k){
    ll->head = reverse_alt_k_nodes_util(ll->head, k);
}

int main() {
    LinkedList* ll = new LinkedList();
    
    // Create Linked list
    ll->insert_at_beg(8);
    ll->insert_at_beg(7);
    ll->insert_at_beg(6);
    ll->insert_at_beg(5);
    ll->insert_at_beg(4);
    ll->insert_at_beg(3);
    ll->insert_at_beg(2);
    ll->insert_at_beg(1);

    cout << "Linked list before reverse: " << endl;
    ll->print_data();
    int k = 3;
    reverse_alt_k_nodes(ll, k);
    cout << "Linked list after reversing in groups of " << k << ": " << endl;
    ll->print_data();
    return 0;
}


